lib/ansible/plugins/inventory/cloudscale.py

```diff
@@ -0,0 +1,185 @@
+            - public_v4
+            - public_v6
+            - private
+            - none
+        default: public_v4
+    api_token:
+        description: cloudscale.ch API token
+        env:
+          - name: CLOUDSCALE_API_TOKEN
+        type: str
+    api_timeout:
+        description: Timeout in seconds for calls to the cloudscale.ch API.
+        default: 30
+        type: int
+'''
+
+EXAMPLES = r'''
+# cloudscale_inventory.yml file in YAML format
+# Example command line: ansible-inventory --list -i cloudscale_inventory.yml
+
+plugin: cloudscale
+'''
+
+from collections import defaultdict
+from json import loads
+from os import environ
+
+from ansible.errors import AnsibleError
+from ansible.module_utils.cloudscale import API_URL
+from ansible.module_utils.urls import open_url
+from ansible.plugins.inventory import BaseInventoryPlugin, Constructable
+
+iface_type_map = {
+    'public_v4': ('public', 4),
+    'public_v6': ('public', 6),
+    'private': ('private', 4),
+    'none': (None, None),
+}
+
+
+class InventoryModule(BaseInventoryPlugin, Constructable):
+
+    NAME = 'cloudscale'
+
+    def _get_server_list(self):
+        # Get list of servers from cloudscale.ch API
+        response = open_url(
+            API_URL + '/servers',
+            headers={'Authorization': 'Bearer %s' % self._token}
+        )
+        return loads(response.read())
+
+    def parse(self, inventory, loader, path, cache=True):
+        super(InventoryModule, self).parse(inventory, loader, path)
+
+        self._read_config_data(path)
+
+        self._token = self.get_option('api_token')
+        if not self._token:
+            raise AnsibleError('Could not find an API token. Set the '
+                               'CLOUDSCALE_API_TOKEN environment variable.')
+
+        inventory_hostname = self.get_option('inventory_hostname')
+        if inventory_hostname not in ('name', 'uuid'):
+            raise AnsibleError('Invalid value for option inventory_hostname: %s'
+                               % inventory_hostname)
+
+        ansible_host = self.get_option('ansible_host')
+        if ansible_host not in iface_type_map:
+            raise AnsibleError('Invalid value for option ansible_host: %s'
+                               % ansible_host)
+
+        # Merge servers with the same name
+        firstpass = defaultdict(list)
+        for server in self._get_server_list():
+            firstpass[server['name']].append(server)
+
+        # Add servers to inventory
+        for name, servers in firstpass.items():
+            if len(servers) == 1 and inventory_hostname == 'name':
+                self.inventory.add_host(name)
+                servers[0]['inventory_hostname'] = name
+            else:
+                # Two servers with the same name exist, create a group
+                # with this name and add the servers by UUID
+                if name not in self.inventory.groups:
+                    self.inventory.add_group(name)
```

Comment:
add_group creates and returns a 'sanitized' group name, to avoid warnings you might want to use `self._sanitize_group_name(name)`

Reply:
ah, I thought this would already made by add_group().

Reply:
I'll do that although I really don't like the fact that this will convert dashes (-) to underscores (_). But this is not specific to this module and if Ansible is really going this way we should probably not allow legacy group names on a new inventory plugin.

Reply:
@resmo it is, but they are not using the results to compare, so same group will never be in inventory.groups (which you don't need to do since add_group already checks this).

Reply:
@gaudenz for those that really want it, there will be global toggle available, you CAN add toggle at plugin level to do same